boot/freeldr/freeldr/arch/uefi/uefidisk.c

```diff
@@ -0,0 +1,589 @@
+{
+    TRACE("UefiDiskOpen: File ID: %d\nPath:%s\n", FileId, Path);
+    DISKCONTEXT* Context;
+    UCHAR DriveNumber;
+    ULONG DrivePartition, SectorSize;
+    ULONGLONG SectorOffset = 0;
+    ULONGLONG SectorCount = 0;
+    ULONG UefiDriveNumber = 0;
+    PARTITION_TABLE_ENTRY PartitionTableEntry;
+
+    if (DiskReadBufferSize == 0)
+    {
+        ERR("DiskOpen(): DiskReadBufferSize is 0, something is wrong.\n");
+        ASSERT(FALSE);
+        return ENOMEM;
+    }
+
+    if (!DissectArcPath(Path, NULL, &DriveNumber, &DrivePartition))
+        return EINVAL;
+
+    TRACE("Opening disk:\n Drive Number: %d\n DrivePartition: %d", DriveNumber, DrivePartition);
+    UefiDriveNumber = DriveNumber - FIRST_BIOS_DISK;
+    GlobalSystemTable->BootServices->HandleProtocol(handles[UefiDriveNumber], &bioGuid, (void **) &bio);
+    SectorSize = bio->Media->BlockSize;
+
+    if (DrivePartition != 0xff && DrivePartition != 0)
+    {
+        if (!DiskGetPartitionEntry(DriveNumber, DrivePartition, &PartitionTableEntry))
+            return EINVAL;
+
+        SectorOffset = PartitionTableEntry.SectorCountBeforePartition;
+        SectorCount = PartitionTableEntry.PartitionSectorCount;
+    }
+    else
+    {
+        GEOMETRY Geometry;
+        if (!MachDiskGetDriveGeometry(DriveNumber, &Geometry))
+            return EINVAL;
+
+        if (SectorSize != Geometry.BytesPerSector)
+        {
+            ERR("SectorSize (%lu) != Geometry.BytesPerSector (%lu), expect problems!\n",
+                SectorSize, Geometry.BytesPerSector);
+        }
+
+        SectorOffset = 0;
+        SectorCount = (ULONGLONG)Geometry.Cylinders * Geometry.Heads * Geometry.Sectors;
+    }
+
+    Context = FrLdrTempAlloc(sizeof(DISKCONTEXT), TAG_HW_DISK_CONTEXT);
+    if (!Context)
+        return ENOMEM;
+
+    Context->DriveNumber = DriveNumber;
+    Context->SectorSize = SectorSize;
+    Context->SectorOffset = SectorOffset;
+    Context->SectorCount = SectorCount;
+    Context->SectorNumber = 0;
+    FsSetDeviceSpecific(*FileId, Context);
+    return ESUCCESS;
+}
+
+static
+ARC_STATUS
+UefiDiskRead(ULONG FileId, VOID *Buffer, ULONG N, ULONG *Count)
+{
+    DISKCONTEXT* Context = FsGetDeviceSpecific(FileId);
+    UCHAR* Ptr = (UCHAR*)Buffer;
+    ULONG Length, TotalSectors, MaxSectors, ReadSectors;
+    ULONGLONG SectorOffset;
+    BOOLEAN ret;
+
+    ASSERT(DiskReadBufferSize > 0);
+
+    TotalSectors = (N + Context->SectorSize - 1) / Context->SectorSize;
+    MaxSectors   = DiskReadBufferSize / Context->SectorSize;
+    SectorOffset = Context->SectorOffset + Context->SectorNumber;
+
+    // If MaxSectors is 0, this will lead to infinite loop.
+    // In release builds assertions are disabled, however we also have sanity checks in DiskOpen()
+    ASSERT(MaxSectors > 0);
+
+    ret = TRUE;
+
+    while (TotalSectors)
+    {
+        ReadSectors = TotalSectors;
+        if (ReadSectors > MaxSectors)
+            ReadSectors = MaxSectors;
```

Comment:
In case this is not coming from already-existing code,
```suggestion
        ReadSectors = min(TotalSectors, MaxSectors);
```